
<launch>
   
   <!-- RGB-D MAPPING VERSION -->
   <!-- WARNING : Database is automatically deleted on each startup -->
   <!--           See "delete_db_on_start" option below... -->
  
  <!-- TF FRAMES -->
  <node pkg="tf" type="static_transform_publisher" name="base_to_camera_tf" 
    args="0.0 0.0 0.0 0.0 0.0 0.0 /base_link /camera_link 100" />
   
  <!-- Odometry -->
  <node pkg="rtabmap" type="visual_odometry" name="visual_odometry" output="screen">
    <remap from="rgb/image"       to="camera/rgb/image_rect_color"/>
    <remap from="depth/image"     to="camera/depth_registered/image_raw"/>
    <remap from="rgb/camera_info" to="camera/depth_registered/camera_info"/>
    
    <param name="Odom/MinInliers" type="string" value="10"/>
	<param name="Odom/InlierDistance" type="string" value="0.01"/>
   
    <param name="frame_id" type="string" value="base_link"/>
  </node>
  
  <group ns="rtabmap">
  
    <!-- Visual SLAM -->
    <!-- args: "delete_db_on_start" and "udebug" -->
    <node name="rtabmap" pkg="rtabmap" type="rtabmap" output="screen" args="--delete_db_on_start">
      <param name="subscribe_depth" type="bool" value="true"/>
      <param name="subscribe_laserScan" type="bool" value="false"/>
	
      <remap from="rgb/image" to="/camera/rgb/image_rect_color"/>
      <remap from="depth/image" to="/camera/depth_registered/image_raw"/>
      <remap from="rgb/camera_info" to="/camera/depth_registered/camera_info"/>
      <remap from="odom" to="/odom"/>
	
      <param name="queue_size" type="int" value="30"/>
    </node>
        
  </group>
  
  <!-- Map assembler for rviz -->
  <node pkg="rtabmap" type="map_assembler" name="map_assembler"  output="screen">
    <remap from="infoEx"       to="rtabmap/infoEx"/>
    <param name="cloud_decimation" type="int" value="4"/>
    <param name="cloud_voxel_size" type="double" value="0.02"/>
    <param name="scan_voxel_size" type="double" value="0.01"/>
  </node>
    
  <node pkg="rviz" type="rviz" name="rviz" args="-d $(find rtabmap)/launch/config/rgbd.rviz"/>
    
  <!-- sync cloud with odometry and voxelize the point cloud (for fast visualization in rviz) -->
  <node pkg="nodelet" type="nodelet" name="standalone_nodelet"  args="manager" output="screen"/>
  <node pkg="nodelet" type="nodelet" name="data_odom_sync" args="load rtabmap/data_odom_sync standalone_nodelet">
    <remap from="rgb/image_in"       to="camera/rgb/image_rect_color"/>
    <remap from="depth/image_in"     to="camera/depth_registered/image_raw"/>
    <remap from="rgb/camera_info_in" to="camera/depth_registered/camera_info"/>
    <remap from="odom_in"             to="odom"/>
    
    <remap from="rgb/image_out"       to="data_odom_sync/image"/>
    <remap from="depth/image_out"     to="data_odom_sync/depth"/>
    <remap from="rgb/camera_info_out" to="data_odom_sync/camera_info"/>
    <remap from="odom_out"            to="odom_sync"/>
    
    <param name="queue_size" type="int" value="30"/>
  </node>
  <node pkg="nodelet" type="nodelet" name="points_xyzrgb" args="load rtabmap/point_cloud_xyzrgb standalone_nodelet">
    <remap from="rgb/image"       to="data_odom_sync/image"/>
    <remap from="depth/image"     to="data_odom_sync/depth"/>
    <remap from="rgb/camera_info" to="data_odom_sync/camera_info"/>
    <remap from="cloud"           to="voxel_cloud" />
    
    <param name="queue_size" type="int" value="10"/>
    <param name="voxel_size" type="double" value="0.01"/>
  </node>

</launch>
